includes/taxonomy.php

```diff
@@ -67,3 +86,81 @@ function register_taxonomy( string $post_type ) {
 		$args
 	);
 }
+
+/**
+ * Determine whether a user can associated shadow terms with posts.
+ *
+ * @return bool True if capable. False if not.
+ */
+function can_associate_posts() : bool {
+	return current_user_can( 'edit_posts' );
+}
+
+/**
+ * Handle the submission of a post association via REST request.
+ *
+ * @param \WP_REST_Request $request The vote submission request.
+ * @return \WP_REST_Response The response data.
+ */
+function handle_rest_associate( \WP_REST_Request $request ) : \WP_REST_Response {
+	$post_id            = (int) $request->get_param( 'postId' );
+	$associated_post_id = (int) $request->get_param( 'associatedPostId' );
+
+	if ( ! API\get_taxonomy_slug( $post_id ) ) {
+		return rest_ensure_response(
+			[
+				'success' => false,
+				'message' => 'This post type is not associated with a shadow taxonomy.',
+				'posts'   => [],
+			]
+		);
+	}
+
+	$post = get_post( $post_id );
+
+	if ( 'publish' !== $post->post_status ) {
+		$associated_posts = API\get_associated_posts( $post_id );
+
+		if ( ! in_array( $associated_post_id, $associated_posts, true ) ) {
+			$associated_posts[] = $associated_post_id;
+			API\update_associated_posts( $post_id, $associated_posts );
+		}
+
+		return rest_ensure_response(
+			[
+				'success' => true,
+				'message' => '',
+				'posts'   => $associated_posts,
+			]
+		);
+	}
+
+	wp_set_object_terms( $associated_post_id, API\get_term_id( $post_id ), API\get_taxonomy_slug( $post_id ) );
+
+	$associated_post  = get_post( $associated_post_id );
+	$associated_posts = new \WP_Query(
+		[
+			'fields'                 => 'ids',
+			'post_type'              => $associated_post->post_type,
```

Comment:
Not sure how critical this is, but it's possible for a post/its shadow term to be associated with posts of different types, right?

Reply:
@philcable Very good point! I think we're okay here for now because we're relying on a list of posts to be returned that matches the freshly associated post. But that's not a guarantee and very much something that needs to be sorted. I'll work on a new issue to describe the intent/objective. Thanks!